### ISSUE TITLE
Rule suggestion: Using rector to fix prophecy typehints

### ISSUE THREAD

**Issue by alexander-schranz** [rectorphp/rector-phpunit#83]
We have a big codebase in @sulu and lots of our testcases are not really analyzeable by phpstan  as there are lot of false typehints for mocks (we have a very very big baseline). We are using `prophecy` for mocking. And there are different false typehints in our codebase:

Variant A (class but didnt define mock):

```php
class MyClassTest extends TestCase {
    /**
     * @var MyClass
     */
    private $my;

    public function setUp(): void
    {
         $this->my = $this->prophesize(MyClass::class);
    }
}
```

Variant B (mock but didnt define class):

```php
class MyClassTest extends TestCase {
    /**
     * @var ObjectProphecy
     */
    private $my;

    public function setUp(): void
    {
         $this->my = $this->prophesize(MyClass::class);
    }
}
```

Variant C (mock or class (is still false but works sometimes)):

```php
class MyClassTest extends TestCase {
    /**
     * @var ObjectProphecy|MyClass
     */
    private $my;

    public function setUp(): void
    {
         $this->my = $this->prophesize(MyClass::class);
    }
}
```

Variant D (nothing defined):

```php
class MyClassTest extends TestCase {
    private $my;

    public function setUp(): void
    {
         $this->my = $this->prophesize(MyClass::class);
    }
}
```

Expected would be ObjectProphecy Generic with the class:

```php
class MyClassTest extends TestCase {
    /**
     * @var ObjectProphecy<MyClass>
     */
    private $my;

    public function setUp(): void
    {
         $this->my = $this->prophesize(MyClass::class);
    }
}
```

So I asked myself if rector could be used to provide a rule to analyze all `$this->prophesize` calls and set the correct type for this classes. Think it would be a great rule which could be shipped with rector via a CodeQuality ruleset. Any hints are welcome how to achieve this :)

**Comment by TomasVotruba** [comment]
Hi, this seems like a great idea :+1: 

We already have a rule that adds types/@var types based on assign, and this seems like exact behavior:
https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#typedpropertyfromassignsrector

It will probably need extension specifically for prophecy, possibly PHPStan extension.

**Comment by alexander-schranz** [comment]
That is a great hint. The `TypedPropertyFromAssignsRector` is a general rule and can I think currently not be configured for a specific property or even method call. Think that would be needed to be adopted.

We use the [https://github.com/Jan0707/phpstan-prophecy](Jan0707/phpstan-prophecy) extension so  phpstan understands the `$this->prophesize`. So I think that extension should be enough to get the result of $this->prophesize.

**Comment by TomasVotruba** [comment]
As stated, the types are based in PHPStan type resolving system, so first you'd have to detect type with propper PHPStan extension.

Another layer would be to add custom rule, that only works with mocks. Because this is resolved, but not valid: 

```php
private MockInterface&RealType $mockedRealType;
```

If you still work on it and have a custom rule that handles these case, I'd be happy to merge it :+1:

**Comment by alexander-schranz** [comment]
Will keep you uptodate. Currently more ideas then time 🙈

**Comment by alexander-schranz** [comment]
fixed in #115
